### ISSUE TITLE
Unable to create pods: not enough IPs available, waiting on Azure CNS to allocate more

### ISSUE THREAD

**Issue by kipusoep** [Azure/azure-container-networking#2999]
**What happened:** 
When the nodepool get updated, pods are rescheduled all at once. This leads to a state in which the pods are not being created, because of IP starvation.
We have Pod CIDR set to 100.112.0.0/12, which means the pool should have 1,048,574 IP addresses available.
But I only see pods with IP address 100.112.1.x - 100.112.5.x. I don't understand why 100.113.x.x isn't used for example?
The error we get:
```
Warning  FailedCreatePodSandBox  2m28s (x587 over 130m)  kubelet  (combined from similar events): Failed to create pod sandbox: rpc error: code = Unknown desc = failed to setup network for sandbox "98f722bb934e891615d307eb75f1300d78fa379f60781ca190f02ca2ca3f6f60": plugin type="azure-vnet" failed (add): IPAM Invoker Add failed with error: failed to add ipam invoker: Failed to get IP address from CNS: AllocateIPConfig failed: not enough IPs available, waiting on Azure CNS to allocate more, IP config request is {[] 98f722bb-eth0 98f722bb934e891615d307eb75f1300d78fa379f60781ca190f02ca2ca3f6f60 [123 34 80 111 10 78 97 109 101 34 58 34 114 111 117 116 105 110 103 45 111 114 99 104 101 115 116 114 97 116 111 114 103 97 116 101 119 97 121 45 56 53 57 102 53 53 55 55 98 53 45 120 50 110 54 55 34 44 34 80 111 100 78 97 109 101 115 112 97 99 101 34 58 34 100 101 102 97 117 108 116 34 125[]  false}
```

**What you expected to happen:**  
There shouldn't be IP starvation at all.

**How to reproduce it:**  
That's a bit hard to describe, sometimes it happens and sometimes it doesn't it seems.

**Orchestrator and Version (e.g. Kubernetes, Docker):**  
K8s version 1.30.3

**Operating System (Linux/Windows):**  
Linux

**Kernel (e.g. `uanme -a` for Linux or `$(Get-ItemProperty -Path "C:\windows\system32\hal.dll").VersionInfo.FileVersion` for Windows):**
Linux version 5.15.0-1071-azure (buildd@lcy02-amd64-063) (gcc (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0, GNU ld (GNU Binutils for Ubuntu) 2.38) #80-Ubuntu SMP Tue Aug 6 19:27:32 UTC 2024

**Anything else we need to know?:**
[Miscellaneous information that will assist in solving the issue.]
Network configuration: Azure CNI Overlay
Pod CIDR: 100.112.0.0/12
Network policy: None

**Comment by rbtr** [comment]
@kipusoep is this a persistent issue or does it resolve itself? Is this cluster using dynamic podsubnet or overlay networking mode? What size/scale is the cluster and workload?
If you see this consistently we want to investigate why - do you have an Azure support case for it?

**Comment by kipusoep** [comment]
> is this a persistent issue or does it resolve itself?
It resolves itself, but we also had this on our production environment and it took quite some time to resolve, so it's a real PITA.
I think this time it took +/- 2 hours.
> Is this cluster using dynamic podsubnet or overlay networking mode?
Not sure? I can only see Azure CNI Overlay for the Network configuration. Where should I look?
> What size/scale is the cluster and workload? Which region?
We have 2 nodepools with 5 nodes in total. +/- 180 pods in total. West Europe is the region.
> If you see this consistently we want to investigate why - do you have an Azure support case for it?
Yes we see this consistenly on all 4 of our clusters (dev/tst/acc/prd). If something triggers nodes to be recreated at the moment, our production workload is in jeopardy.
I'm afraid I am not able to create an Azure support case...

**Comment by rbtr** [comment]
> I think this time it took +/- 2 hours.

that is unexpected, seems like something is indeed wrong. especially at that relatively small scale.

> I'm afraid I am not able to create an Azure support case...

we need to be able to identify your cluster to do any investigation, maybe you could email me the cluster FQDN? my full name @microsoft

**Comment by kipusoep** [comment]
Thanks Evan, I've sent you an email just now. Can you confirm you received it?

**Comment by rbtr** [comment]
I did

**Comment by rbtr** [comment]
Hey @kipusoep, I had a look through the logs and I don't see anything obviously wrong here. Transient Pod startup delays may occur if Nodes are restarted or the Node pool is scaled frequently as the underlying network is provisioned. This may be worse if the cluster is under-provisioned or overloaded. 
Since the Pods do end up with functional networking, this seems unlikely to be an issue in the networking components themselves. Consistently 2h delays would be unexpected, but on the order of minutes is normal.
A change which may resolve this issue for you is already rolling out - the Nodes will stay in NetworkNotReady until they actually have IPs provisioned that are ready to provide to Pods, and then the Node will become Ready for Pod scheduling. Until then no Pods should be scheduled there.

Unfortunately, without a support ticket there's not much more investigation I can do here. The best next steps would be to open an Azure Support request so that the AKS team can take a look and re-involve container networking if necessary.

**Comment by kipusoep** [comment]
Hey @rbtr, as mentioned the logs that I provided were not originating from the same situation.
I still need to reproduce the issue and harvest some logs.

Having Nodes waiting to be provisioned with enough IPs would help, but that means new nodes would be unschedulable for 2 hours and that's still unacceptable.

I must admit I'm a bit confused about the way this report goes; first you want to help and now all of a sudden I need an Azure support request.
In any case I will need to provide logs I guess, so I will still go and reproduce the issue, harvest some logs and then come back here. If you still don't see anything out of the ordinary I will move on to our CSP to create a support ticket (but I'd much like to prevent that, as it's a cumbersome process).

**Comment by rbtr** [comment]
I still want to help; I'm telling you what you need to do so that we can continue to help.

This repo is for opensource collaboration on azure-cni and azure-cns, and from what you have provided so far, those components appear to be operating normally. The behavior you're describing is not isolated to these components, it's with [AKS](https://github.com/Azure/AKS) as a system.

What you're asking for is production support for your AKS cluster(s). That needs to go through Azure support for several reasons, such as that they will collect logs and data from your environment that I cannot directly access without a real prod support request, try to repro, consolidate all the investigation into a single place, link related reports for tracking, and manage things like communicating impact and an outage to affected users like yourself if this turned out to be a widespread issue. 

That's the reason why one of the first things I asked was if you had a support ticket:

> ...[do you have an Azure support case for it?](https://github.com/Azure/azure-container-networking/issues/2999#issuecomment-2338336975)

**Comment by kipusoep** [comment]
So I just ran into this again by accident and I was able to fetch the Azure CNS pod log as well.

The log of the pod itself:
```
rpc error: code = Unknown desc = failed to setup network for sandbox "4ab223aa23a33d10a539a1077f0b3f900a09e75c5659f6464daef6a60d3007bd": plugin type="azure-vnet" failed (add): IPAM Invoker Add failed with error: failed to add ipam invoker: Failed to get IP address from CNS: AllocateIPConfig failed: not enough IPs available for 09265823-4093-4882-8ff7-a0812ef4edb2, waiting on Azure CNS to allocate more with NC Status: , IP config request is {[] 4ab223aa-eth0 4ab223aa23a33d10a539a1077f0b3f900a09e75c5659f6464daef6a60d3007bd [123 34 80 111 100 78 97 109 101 34 58 34 102 114 111 110 116 101 110 100 45 97 110 116 119 101 114 112 101 110 45 53 102 98 52 54 100 102 54 54 53 45 53 108 53 122 55 34 44 34 80 11 100 78 97 109 101 115 112 97 99 101 34 58 34 100 101 102 97 117 108 116 34 125[]  false false []}
```

And the log from the Azure CNS pod:
```
2024/10/18 11:37:59 [1] [azure-cnsrequestIPConfigsHandler] Code:FailedToAllocateIpConfig, {DesiredIPAddresses:[] PodInterfaceID:75c0b533-eth0 InfraContainerID:75c0b5330df1417b80d5774cd44d35ad3a322976c235b4551f388dac94e86591 OrchestratorContext:[123 34 80 111 100 78 97 109 101 34 58 34 102 114 111 110 116 101 110 100 45 97 110 116 119 101 114 112 101 110 45 53 102 98 52 54 100 102 54 54 53 45 53 108 53 122 55 34 44 34 80 111 100 78 97 109 101 115 112 97 99 101 34 58 34 100 101 102 97 117 108 116 34 125] Ifname: SecondaryInterfacesExist:false BackendInterfaceExist:false BackendInterfaceMacAddresses:[]}, &{PodIPInfo:[{PodIPConfig:{IPAddress: PrefixLength:0} NetworkContainerPrimaryIPConfig:{IPSubnet:{IPAddress: PrefixLength:0} DNSServers:[] GatewayIPAddress:} HostPrimaryIPInfo:{Gateway: PrimaryIP: Subnet:} NICType: InterfaceName: MacAddress: SkipDefaultRoutes:false Routes:[] PnPID:} {PodIPConfig:{IPAddress: PrefixLength:0} NetworkContainerPrimaryIPConfig:{IPSubnet:{IPAddress: PrefixLength:0} DNSServers:[] GatewayIPAddress:} HostPrimaryIPInfo:{Gateway: PrimaryIP: Subnet:} NICType: InterfaceName: MacAddress: SkipDefaultRoutes:false Routes:[] PnPID:}] Response:{ReturnCode:FailedToAllocateIpConfig Message:AllocateIPConfig failed: not enough IPs available for 09265823-4093-4882-8ff7-a0812ef4edb2, waiting on Azure CNS to allocate more with NC Status: , IP config request is {[] 75c0b533-eth0 75c0b5330df1417b80d5774cd44d35ad3a322976c235b4551f388dac94e86591 [123 34 80 111 100 78 97 109 101 34 58 34 102 114 111 110 116 101 110 100 45 97 110 116 119 101 114 112 101 110 45 53 102 98 52 54 100 102 54 54 53 45 53 108 53 122 55 34 44 34 80 111 100 78 97 109 101 115 112 97 99 101 34 58 34 100 101 102 97 117 108 116 34 125]  false false []}}}.
```
